azure-iot-device/azure/iot/device/provisioning/pipeline/pipeline_stages_provisioning.py

```diff
@@ -44,3 +58,388 @@ def _run_op(self, op):
+                                        status_code=success_status_code
+                                    )
+                                )
+                        else:
+                            error = exceptions.ServiceError(
+                                "Query Status Operation encountered an invalid registration status {status} with a status code of {status_code}".format(
+                                    status=registration_status, status_code=success_status_code
+                                )
+                            )
+
+                        query_status_op.complete(error=error)
+
+                else:
+                    query_status_op.complete(error=error)
+
+            self.send_op_down(
+                pipeline_ops_base.RequestAndResponseOperation(
+                    request_type=constant.QUERY,
+                    method="GET",
+                    resource_location="/",
+                    query_params={"operation_id": op.operation_id},
+                    request_body=op.request_payload,
+                    callback=on_query_response,
+                )
+            )
+
+        else:
+            super(PollingStatusStage, self)._run_op(op)
+
+
+class RegistrationStage(CommonProvisioningStage):
+    """
+    This is the first stage that decides converts a registration request
+    into a normal request and response operation.
+    Upon the receipt of the response this stage decides whether
+    to send another registration request or send a query request.
+    Depending on the status and result of the response
+    this stage may also complete the registration process.
+    """
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _run_op(self, operation):
+        if isinstance(operation, pipeline_ops_provisioning.SendRegistrationRequestOperation):
+            initial_register_op = operation
+
+            def on_registration_response(op, error):
+                logger.debug(
+                    "{stage_name}({op_name}): Received response with status code {status_code} for SendRegistrationRequestOperation".format(
+                        stage_name=self.name, op_name=op.name, status_code=op.status_code
+                    )
+                )
+
+                # This could be an error that has been reported to this stage Or this
+                # could be an error because the service responded with status code >=300
+                error = self._get_error(op, REGISTER, error=error)
+
+                if not error:
+                    success_status_code = op.status_code
+                    retry_interval = (
+                        int(op.retry_after, 10)
+                        if op.retry_after is not None
+                        else constant.DEFAULT_POLLING_INTERVAL
+                    )
+                    decoded_response = self._decode_response(op)
```

Comment:
>decoded_response = self._decode_response(op) [](start = 20, length = 44)

This is only relevant in the success block

Reply:
i am not understanding what you mean. 

---
In reply to: [365408409](https://github.com/Azure/azure-iot-sdk-python/pull/432#discussion_r365408409) [](ancestors = 365408409)

Reply:
Never mind, while decoded response itself is only elsewhere used in the success block, operation id is derived from it, and it is used in a couple different places, thus must be defined here.

---
In reply to: [366010391](https://github.com/Azure/azure-iot-sdk-python/pull/432#discussion_r366010391) [](ancestors = 366010391,365408409)